Cython/Compiler/Code.py

```diff
@@ -1556,6 +1600,9 @@ def initialize_main_c_code(self):
         code.putln("")
         code.putln("/* --- Runtime support code --- */")
 
+    def register_part(self, name, code):
+        self.parts[name] = code.insertion_point()
```

Comment:
Rather than injecting new code layout parts into the file at random points, they should always exist in the `code_layout` list and just stay empty if they are unused.

I think all of the `__Pyx_ExportFunction()` calls, including the already existing ones, can go into a separate C function (maybe `__Pyx_ExportModuleFunctions`) with a separate code layout part. The new shared functions can then just add code to that part.

The same applies to the function importing. Just add a C function (and layout section) for that as well, leave it empty if unused, fill it with function imports (and maybe also the shared module import) if used. Inside of such a function, you're free to use your own local variables and don't need to resort to allocating temps etc.